notebook/services/kernels/kernelmanager.py

```diff
@@ -99,6 +101,47 @@ def shutdown_kernel(self, kernel_id, now=False):
         self._check_kernel_id(kernel_id)
         super(MappingKernelManager, self).shutdown_kernel(kernel_id, now=now)
 
+    def restart_kernel(self, kernel_id):
+        """Restart a kernel by kernel_id"""
+        self._check_kernel_id(kernel_id)
+        super(MappingKernelManager, self).restart_kernel(kernel_id)
+        kernel = self.get_kernel(kernel_id)
```

Comment:
All connections in this process remain valid across restart, both zmq and websocket. They do not need to be closed.